### PR TITLE
Implemented Python to C++ Server Communication

### PR DESCRIPTION
**Pull request by andrewmzhang**
Added code to paramserver, error task, and cirrus python so that users can grab time-error and num lambdas.

Also added a .gitignore to cirrus python

### CODE REVIEW

File: python/frontend/cirrus/cirrus/messenger.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    clientsocket.sendto(GET_LAST_TIME_ERROR, (ip, port))
+    s = clientsocket.recv(128)
+    return struct.unpack("dd", s)
+
+def get_all_time_error(ip="127.0.0.1", port=1338):
+    clientsocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    clientsocket.sendto(GET_ALL_TIME_ERROR, (ip, port))
+    s = clientsocket.recv(32)
+    return "NOT IMPLEMENTED"
+if __name__ == "__main__":
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/messenger.py]: This doesn't seem well formatted. Also can you add a comment explaining what this script does?

File: src/ErrorSparseTask.cpp [hunk 2]
```diff
@@ -27,7 +29,57 @@ std::unique_ptr<CirrusModel> get_model(const Configuration& config,
   return psi->get_full_model(use_col_filtering);
 }
 
+void ErrorSparseTask::error_response() {
+  int fd;
+  if ((fd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
+    std::cout << "SOCKET FAILED"
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: Why don't you throw an exception?

File: src/ErrorSparseTask.cpp [hunk 3]
```diff
@@ -27,7 +29,57 @@ std::unique_ptr<CirrusModel> get_model(const Configuration& config,
...
+                      (struct sockaddr*) &remaddr, &addrlen);
+    if (length < 0) {
+      std::cout << "Failed to read" << std::endl;
+      continue;
+    }
+    printf("Received: %d bytes\n", length);
+    std::cout << "Received: " << operation << std::endl;
+
+    if (operation == GET_LAST_TIME_ERROR) {
+      double time_error[2];
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: double time_error[2] = {last_time, last_error};

File: src/ErrorSparseTask.cpp [hunk 4]
```diff
@@ -27,7 +29,57 @@ std::unique_ptr<CirrusModel> get_model(const Configuration& config,
...
+    return;
+  }
+
+  struct sockaddr_in serveraddr;
+  serveraddr.sin_family = AF_INET;
+  serveraddr.sin_addr.s_addr = INADDR_ANY;
+  serveraddr.sin_port = htons(1338);
+  std::memset(serveraddr.sin_zero, 0, sizeof(serveraddr.sin_zero));
+
+  if (bind(fd, (struct sockaddr*) &serveraddr, sizeof(serveraddr)) < 0) {
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: Can you use C++-style casts here and below?

File: src/ErrorSparseTask.cpp [hunk 5]
```diff
@@ -27,7 +29,57 @@ std::unique_ptr<CirrusModel> get_model(const Configuration& config,
...
+    }
+    printf("Received: %d bytes\n", length);
+    std::cout << "Received: " << operation << std::endl;
+
+    if (operation == GET_LAST_TIME_ERROR) {
+      double time_error[2];
+      time_error[0] = last_time;
+      time_error[1] = last_error;
+
+      sendto(fd, time_error, 2 * sizeof(double), 0, (struct sockaddr*) &remaddr,
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: Always check return values (here and below).

File: src/Tasks.h [hunk 6]
```diff
@@ -153,8 +153,11 @@ class ErrorSparseTask : public MLTask {
           nworkers, worker_id, ps_ip, ps_port)
   {}
     void run(const Configuration& config);
+    void error_response();
 
-  private:
+   private:
+    double last_time = 0.0;
```
Comment (jcarreira) [inline src/Tasks.h]: Comments in every class attribute.

File: python/frontend/cirrus/cirrus/messenger.py (hunk 1 above)
Comment (andrewmzhang) [reply python/frontend/cirrus/cirrus/messenger.py]: I intended for that to be something that collects the entire history of time vs error from an Error Task. However there is currently no code on the Error Task that allows for this, which is why the frontend isn't implemented. I will remove this for now.

File: src/PSSparseServerTask.cpp [hunk 7]
```diff
@@ -399,7 +399,10 @@ void PSSparseServerTask::gradient_f() {
                 << std::endl;
 #endif
       task_to_status[data[0]] = data[1];
-    
+
+    } else if (operation == GET_NUM_CONNS) {
+      std::cout << "Retrieve num connections: " << num_connections << std::endl;
+      send(sock, &num_connections, sizeof(uint32_t), 0);
```
Comment (jcarreira) [inline src/PSSparseServerTask.cpp]: I think this return value should be checked.

**Pull request merged**